README.md
# Version 0.15

# Project Goal
This project provides a initial pelican configs and theme files to offer a painless experience to publish jupyter notebooks as pelican blogs into gitpages.

Although pelican has ipynb plugin to convert .ipynb file into blog html, many adaptations have to be made to support jupyter notebook bloggin.

primary adaptations

* .ipynb files seldomly exist alone, they usually work with related static files such as test data, images, etc. They are better organized in  one same directorys. Pelicanconf.py is configured so that pelican can detect .ipynb files in directorys and the links are not broken after they are converted into html
* In .ipynb generated html, contents are nested deeply in <div> elements so some javascript plugins such as tocify.js cannot work properly. So scripts are added in article template to properly extract content from the nested <div> block
* 'Clean-blog' theme in pelican provides a clean, good looking but provide too little useful component common to blogs, such as
    * Duoshuo Comments (disqus is blocked for chinese users)
    * Responsive sidebar toc menu
    * Menus for category, tag, archives
    * Wechat award image

# Demonstration
My jupyter notebooks are kept in [here](https://github.com/junjiecai/jupyter_labs/tree/master/exolution)　and you can check what they look like after they are published into blogs [here](https://junjiecai.github.io)


# Perquisites
If you have no idea about pelican and gitpages, please read one of these articles first.

* [Building a data science portfolio: Making a data science blog](https://www.dataquest.io/blog/how-to-setup-a-data-science-blog/)
* [my blog on the same topic (Chinese)](http://junjiecai.github.io/posts/2016/Aug/10/blog_with_jupyter/)


# How to use this project
After cloning the repo to your local machine, finish the following tasks


## create virtualenv (Recommended)
Install virtualenv if you don't have it
```
pip install virtualenv
```
Change directory into the project folder and create a virutalenv for it, python3.5 or higher is recommended

```
virtualenv -p /usr/bin/python3.5 venv
```

Activate the virtualenv

```
source venv/bin/activate
```

## install requirements
create a requirements.txt and put in the following lines
```
markdown
pelican==3.7.1
jupyter
ipython
nbconvert
beautifulsoup4
matplotlib
ghp-import
```

Then run

```
pip install -r requirements.txt
```


## Register DUOSHUO (Optional)
DUOSHUO provides comments module for your blogs. If you want it, register a DUOSHUO service [here](http://duoshuo.com/) for your blog and take note on your DUOSHUO site name. (looks like 'XXXX.duoshuo.com')


## Register Google Analytics(Optional)
Register a google analytics service [here](https://analytics.google.com/) for your blog site and take note on the google analytics id (looks like 'UA-xxxxxxxx-x')


## Create a customs.py
create a customs.py file (same folder with pelicanconf.py) and add the following content.


### Required Settings
```
AUTHOR = 'YourName'
SITENAME = "YourSiteName"
SITEURL = "YourSiteUrl" # For example, https://junjiecai.github.io 
ARTICLE_PATHS = ['articles'] # your article paths, relative to PATH in pelicanconf.py. See examples below for more detail.
```

### Optional Settings
```
GOOGLE_ANALYTICS = 'UA-xxxxxxxx-x' # (OPTIONAL) Enable google_analytics for your blog.

WECHAT_PAYMENT_IMAGE = 'award.jpeg' # (OPTIONAL) Enable wechat payment qrc at the bottom of each article. This is the filename for your wechat payment qrc image which should be put int theme/static/customs/ folder

HEAD_COVER_IMAGE = 'bg-image.jpg' # (OPTIONAL) Replace the default header cover image if you wish. This is the filename for your header cover   image which should be put int theme/static/customs/ folder

DUOSHUO_SITENAME = "your duoshuo site name" # (OPTIONAL) Enable DUOSHUO commenting. If your site name is 'XXXX.duoshuo.com', just fill in XXXX here

JUPYTER_BASE = 'base url' # (OPTIONAL) Provide this to enable auto source code mapping. See explaination below for more detail.
```


### More explanation about settings
#### ARTICLE_PATHS
For example, here we have files structured like this

```
content
|   articles
    |   jupyter_labs
        |   labs_A
            |   article_0
                |   article0.ipynb
                |   article0.ipynb-meta
            |   article_1
                |   article1.ipynb
                |   article1.ipynb-meta
        |   labs_B

    |   lifes
        |   article_2
            |   article2.ipynb
            |   article2.ipynb-meta
|   pages
```

And we only want to publish contents in directory labs_A and lifes to blog posts, then set ARTICLE_PATHS as
ARTICLE_PATHS = ['articles/jupyter_labs/labs_A','articles/lifes']



#### JUPYTER_BASE
Sometimes, users wish to push the .ipynb files and related resource files into another repository on github. It's convient to automatically map blogs with urls pointing their source files in github. JUPYTER_BASE is used for this situation.

When JUPYTER_BASE is set properly. A automatically generated paragraph is inserted at the beginning of each blog after publishing.

For example, I have files structured like this

```
content
|   articles
    |   jupyter_labs
        |   exolution
            |   0000_blog_with_jupyter # folder name containing .ipynb file
                |   0000_blog_with_jupyter.ipynb
                |   0000_blog_with_jupyter.ipynb-meta
|   pages
```

After I push jupyter_labs directory (contains the source files) into my repository (https://github.com/junjiecai/jupyter_labs), the url pointing to 0000_blog_with_jupyter.ipynb is 

https://github.com/junjiecai/jupyter_labs/tree/master/exolution/0000_blog_with_jupyter

This url can be seperated into two parts

* 'https://github.com/junjiecai/jupyter_labs/tree/master/exolution/', which is invariant between articles. This is the value of JUPYTER_BASE
* 0000_blog_with_jupyter, folder name containing the .ipynb file and related resources. This is automatically extracted and appended into JUPYTER_BASE as the full url.



## Create content folder
Create a folder named with 'content' in the project folder


## Write articles
1. Create a folder (it must be subdirectory of paths defined in ARTICLE_PATHS), you can create one folder for each jupyter notebook lab.
2. Create a .ipynb file and write down the content you wish
3. Create a .ipynb-meta file to provide meta-data about your blog, such as id, title, description .etc, the .ipynb-meta file must has same filename with the .ipynb file

Then you should end up with the directories and files like this.
```
content
|   articles
    |   pytest_intro
        |   pytest_intro.ipynb
        |   pytest_intro.ipynb-meta

```

Within the pytest_intro.ipynb-meta, it should looks like this.
```
Title: Pytest Introduction

Date: 2016-08-15 21:05
Modified: 2016-08-15 21:05
Tags: python, testing
Category:jupyter_labs
Author: Exolution
Summary: Introduction for pytest
Id:0001
Slug: pytest_intro
Lang:ch
```

'Id' is a number assigned to each blog and should be kept unqiue among all blogs, or comments system like DUOSHUO won't work properly.

'Lang' is used to add bilingual language support of the articles. See explanation later.

Id, Slug, Lang can be automatically extracted from filenames of .ipynb if they are named with the pattern like "<id>_<slug>_<lang>.ipynb", for example, "0001_pytest_intro_ch.ipynb" defines id = '0001', slug = 'pytest_intro' and lang = 'ch'. Then you don't have to provide them in .ipynb-meta files.

'Category' be extracted from the first level path after content path for each article. For example, given the following structure
```
content
|   articles
    |   pytest_intro
        |   0001_pytest_intro_ch.ipynb

```
pytest_intro will have a category of 'article', unless you defined another value in .ipynb-meta and overwrite it.



## Add translations for articles
If you want to add english translation for an article. Write another .ipynb in English and set it with the same slug but lang equals to 'en'. You can set them in .ipynb-meta file or take advantage of the auto meda data extraction if you name your files with proper pattern. For example.
```
content
|   articles
    |   pytest_intro
        |   0001_pytest_intro_ch.ipynb
        |   0001_pytest_intro_ch.ipynb-meta
        |   0001_pytest_intro_en.ipynb
        |   0001_pytest_intro_en.ipynb-meta
```

When the html is generated, you will see a tranlation link at the header of the article.


## Add pages
You can add pages such as aboutme page, links page in content/pages. They are created in the same way with articles. But links for pages appear in the navigation menu instead of article list.


## Add your gitpages remote
```
git remote add blog git@github.com:[username]/[username].github.io.git
```
Replace [username] with your github username

## Generate html and push to gitpages
Run these three commands and enjoy your blogs.
```
pelican
ghp-import output -b master
git push blog master
```

** Don't forget to delete .ipynb_checkpoints files before run ```pelican``` command or else you will receive a lot of unpleasant error messages**



CHANGELOG.md
# Change Log
## 0.14
### improvement
* add chinese README

## 0.13.1
### improvement
* add lang and category infomation to DUOSHUO data-thread-key and DUOSHUO data-title

## 0.13
### New Features
* Category can be extracted from the first level path after content path for each article
* Side bar toc detect h1~t4 now
* Generated Site can be bilingual now (support Chinese and English).
* Hide ```div.prompt``` in generated HMTL for better visual effect.


## 0.12.1
### New Features
* Make wechat payment, DUOSHUO comments optional components and they can be configed in custom.py.
* Users can change header cover image through customs.py

### Refactoring
* html template for wechat payment, source code mapping are moved into indivisual template files.

### Bug fixes
* Directory name generated from date information always use english now.

## 0.11
### New Features
* Support JUPYTER_BASE config in customs.py to auto-append a paragraph in each article, linking to url containing related .ipybn file and other resources.


## 0.10.2
### Bug fixes
* Fix several text errors in readme.
* Fix tags page "404 not found" error.


## 0.10.1
### New Features
* Auto extract id and slug from filename if it's named with convention 'xxxx_slug'. They can be overwritten in the .ipynb-meta file.

### Bug fixes
* Fix several text errors in readme.
* Fix errors in pelicanconf.py








pelicanconf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- #
from __future__ import unicode_literals

# Don't touch this

import sys
sys.path.append('')
import locale
locale.setlocale(locale.LC_ALL, "en_US.UTF-8")

from customs import *


MENUITEMS = [
            ('分类','categories.html'),
            ('标签','tags.html'),
            ('归档','archives.html')
        ]

I18N_SUBSITES['en']['MENUITEMS'] = [
            ('Categories','categories.html'),
            ('Tags','tags.html'),
            ('Archives','archives.html')
        ]


DUOSHUO_PREFIX = SITEURL.split(':')[1]
DIRECT_TEMPLATES = ['index', 'categories', 'authors', 'archives','tags']

YEAR_ARCHIVE_SAVE_AS = 'posts/{date:%Y}/index.html'
MONTH_ARCHIVE_SAVE_AS = 'posts/{date:%Y}/{date:%b}/index.html'

DATE_FORMATS = {
    'en': '%d/%m/%Y',
    'ch': '%Y-%m-%d'
}


PATH = 'content/'

PAGE_PATHS = ['pages']
DISPLAY_PAGES_ON_MENU = True


FILENAME_METADATA = '(?P<id>\d+)_(?P<slug>.*?)_(?P<lang>ch|en)'

PATH_METADATA = "articles/(?P<category>.*?)/(.*/)(?P<folder>.*)(/.*)"


ARTICLE_URL = 'posts/{date:%Y}/{date:%b}/{date:%d}/{slug}/'
ARTICLE_SAVE_AS = 'posts/{date:%Y}/{date:%b}/{date:%d}/{slug}/index.html'

ARTICLE_ORDER_BY = 'reversed-date'

PAGE_URL = 'pages/{slug}/'
PAGE_SAVE_AS = 'pages/{slug}/index.html'

LOAD_CONTENT_CACHE = False

TIMEZONE = 'Asia/Shanghai'

LOCALE = 'en_US.utf8'

DEFAULT_LANG = 'ch'
STATIC_LANG_URL = 'ch'
STATIC_LANG_SAVE_AS = 'ch'

# Feed generation is usually not desired when developing
FEED_ALL_ATOM = None
CATEGORY_FEED_ATOM = None
TRANSLATION_FEED_ATOM = None
AUTHOR_FEED_ATOM = None
AUTHOR_FEED_RSS = None

# Blogroll
LINKS = (('Pelican', 'http://getpelican.com/'),
         ('Python.org', 'http://python.org/'),
         ('Jinja2', 'http://jinja.pocoo.org/'),
         ('You can modify those links in your config file', '#'),)



# Social widget
SOCIAL = (('You can add links in your config file', '#'),
          ('Another social link', '#'),)

DEFAULT_PAGINATION = 10

# Uncomment following line if you want document-relative URLs when developing
#RELATIVE_URLS = True


MARKUP = ('md', 'ipynb')

PLUGIN_PATHS = ['plugins']
PLUGINS = ['ipynb.markup','i18n_subsites']
IPYNB_USE_METACELL = True
IGNORE_FILES = ['.ipynb_checkpoints']


#,'', 'extract_toc','neighbors','pdf'

## Theme Config for pelican-clean-blog
THEME = "themes"

IPYNB_USE_META_SUMMARY = True

HEADER_COVER = '../theme/images/home-bg.jpg'

COLOR_SCHEME_CSS = 'github.css'


customs.py



AUTHOR = 'Exolution'
SITENAME = "Exolution的奇妙冒险"
BASEURL='https://junjiecai.github.io'
SITEURL='https://junjiecai.github.io'
GITHUB_URL = 'https://github.com/junjiecai/jupyter_labs'
GOOGLE_ANALYTICS = 'UA-89522436-1'
ARTICLE_PATHS = ['articles/jupyter_labs/exolution']
JUPYTER_BASE = 'https://github.com/junjiecai/jupyter_labs/tree/master/exolution'
WECHAT_PAYMENT_IMAGE = 'award.jpeg'
HEADER_COVER_IMAGE = 'bg_image.jpg'


I18N_SUBSITES = {
    'en': {
        'SITENAME':"Exolution's Bizarre Adventure"
        }
    }


from os import listdir, getcwd, makedirs, remove
from shutil import rmtree, copyfile
from os.path import join, isdir, exists


cwd = getcwd()

# seems useless
STATIC_PATHS = []
for artical_path in ARTICLE_PATHS:
    STATIC_PATHS += [join(artical_path, blog_path) for blog_path in listdir(join(cwd, 'content', artical_path))]


month_mapping = {
    '01': 'Jan',
    '02': 'Feb',
    '03': 'Mar',
    '04': 'Apr',
    '05': 'May',
    '06': 'Jun',
    '07': 'Jul',
    '08': 'Aug',
    '09': 'Sep',
    '10': 'Oct',
    '11': 'Nov',
    '12': 'Dec'
}

def get_file_info(full_path):
    file_name = full_path.split('/')[-1].split('.')[0]
    folder_name = file_name[:-3]
    lang_name = file_name[-2:]

    year = None

    with open(full_path) as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            if line.startswith('Date'):
                time_string = line.split(':')[1].strip()
                year, month, date = time_string[:10].split('-')


    return year, month, date, folder_name, lang_name





def search_posts():
    data = []
    for artical_path_relative in ARTICLE_PATHS:
        artical_path_absolute = join(cwd, 'content', artical_path_relative)
        for post_folder_relative in listdir(artical_path_absolute):
            post_folder_absolute = join(cwd, 'content', artical_path_relative, post_folder_relative)
            if isdir(post_folder_absolute):
                image_paths_absolute = []
                post_paths_absolute = []
                for filename in listdir(post_folder_absolute):
                    file_path_absolute = join(post_folder_absolute, filename)
                    if filename.endswith('.ipynb_checkpoints'):
                        rmtree(file_path_absolute)
                    elif filename.endswith('.DS_Store'):
                        remove(file_path_absolute)
                    elif filename.endswith('.jpeg') or filename.endswith('.png'):
                        image_paths_absolute.append(file_path_absolute)
                    elif filename.endswith('nbdata'):
                        year, month, day, post_folder_relative, lang_name = get_file_info(file_path_absolute)
                        month = month_mapping[month]
                        if lang_name == 'ch':
                            target_folder = join(cwd, 'output', 'posts', year, month, day, post_folder_relative[5:])
                        else:
                            target_folder = join(cwd, 'output', 'en','posts', year, month, day, post_folder_relative[5:])
                        post_paths_absolute.append(target_folder)
                    else:
                        pass
                data.append((post_paths_absolute, image_paths_absolute))

    return data


# copy static image into output folder
def copy_image(data):
    for post_data, images in data:
        for target_folder in post_data:
            for image in images:
                src = image
                if not exists(target_folder):
                    makedirs(target_folder)
                target = join(target_folder, image.split('/')[-1])
                copyfile(src, target)


data = search_posts()
copy_image(data)





